### ISSUE TITLE
OpenCV 4.8.1 Error: Bad new number of rows in cv::Mat::reshape

### ISSUE THREAD

**Issue by cesarpgouveia** [opencv/opencv#24418]
### System Information

OpenCV version: 4.8.1
Operating System / Platform: Windows 10
OpenVINO version: 2022.3.1 (2022.3.1.9227.cf2c7da5689_x86_64)

Flags used to build OpenCV with OpenVINO:

`cmake .. -DBUILD_opencv_apps=OFF -DBUILD_opencv_aruco=ON -DBUILD_opencv_bgsegm=OFF -DBUILD_opencv_bioinspired=OFF -DBUILD_opencv_calib3d=ON -DBUILD_opencv_ccalib=ON -DBUILD_opencv_core=ON -DBUILD_opencv_datasets=OFF -DBUILD_opencv_dnn=ON -DBUILD_opencv_dnn_objdetect=OFF -DBUILD_opencv_dnn_superres=OFF -DBUILD_opencv_dpm=OFF -DBUILD_opencv_face=OFF -DBUILD_opencv_features2d=ON -DBUILD_opencv_flann=ON -DBUILD_opencv_fuzzy=OFF -DBUILD_opencv_gapi=OFF -DBUILD_opencv_hfs=OFF -DBUILD_opencv_highgui=ON -DBUILD_opencv_img_hash=ON -DBUILD_opencv_imgcodecs=ON -DBUILD_opencv_imgproc=ON -DBUILD_opencv_intensity_transform=OFF -DBUILD_java_bindings_generator=OFF -DBUILD_opencv_js=OFF -DBUILD_opencv_line_descriptor=OFF -DBUILD_opencv_mcc=OFF -DBUILD_opencv_ml=ON -DBUILD_opencv_objc_bindings_generator=OFF -DBUILD_opencv_objdetect=ON -DBUILD_opencv_optflow=OFF -DBUILD_opencv_phase_unwrapping=OFF -DBUILD_opencv_photo=ON -DBUILD_opencv_plot=ON -DBUILD_opencv_python3=OFF -DBUILD_opencv_python_bindings_generator=OFF -DBUILD_opencv_python_tests=OFF -DBUILD_opencv_quality=OFF -DBUILD_opencv_rapid=OFF -DBUILD_opencv_reg=OFF -DBUILD_opencv_rgdb=OFF -DBUILD_opencv_saliency=OFF -DBUILD_opencv_shape=OFF -DBUILD_opencv_stereo=OFF -DBUILD_opencv_stitching=OFF -DBUILD_opencv_structured_light=OFF -DBUILD_opencv_superres=OFF -DBUILD_opencv_surface_matching=OFF -DBUILD_opencv_text=OFF -DBUILD_opencv_tracking=ON -DBUILD_opencv_ts=OFF -DBUILD_opencv_video=ON -DBUILD_opencv_videoio=ON -DBUILD_opencv_videostab=OFF -DBUILD_opencv_world=ON -DBUILD_opencv_xfeatures2d=ON -DBUILD_opencv_ximgproc=ON -DBUILD_opencv_xobjdetect=ON -DBUILD_opencv_xphoto=ON -DCMAKE_CONFIGURATION_TYPES=Release -DCMAKE_GENERATOR_PLATFORM=x64 -T v142 -DWITH_FFMPEG=ON -DWITH_INF_ENGINE=ON -DENABLE_CXX11=ON -DOPENCV_ENABLE_NONFREE=OFF -DBUILD_JAVA=OFF -DWITH_MSMF=OFF -DWITH_MSMF_DXVA=OFF -DOPENCV_EXTRA_MODULES_PATH="C:/Users/cesar.gouveia/Projects/OpenCV-Package/opencv_contrib/modules" -DBUILD_ZLIB=ON -DBUILD_opencv_hdf=OFF`

### Detailed description

Hi everyone,

I'm encountering a problem with OpenCV 4.8.1, and I'm hoping to get some insights or solutions from the community. I've been running a computer vision model that works perfectly in OpenCV 4.7.0. However, when I tried to migrate to OpenCV 4.8.1 and run the same model on the CPU using both the Inference Engine and OpenCV, I encountered an error that I can't seem to resolve.

The error message I'm getting is as follows:

```
OpenCV: terminate handler is called! The last OpenCV error is:
OpenCV(4.8.1) Error: One of the arguments' values is out of range (Bad new number of rows) in cv::Mat::reshape, file C:\Users\cesar.gouveia\Projects\OpenCV-Package\opencv_mirror\modules\core\src\matrix.cpp, line 1177.
In simpler terms, this error appears to be related to an issue with the number of rows specified for a matrix in OpenCV's core code. This problem leads to an unexpected program termination.
```

The same code and model work smoothly with OpenCV 4.7.0. I would greatly appreciate any insights, advice, or potential solutions from the community. Have any of you encountered a similar issue when transitioning to OpenCV 4.8.1?

The model triggering the error: 
[model1_withFlatten.zip](https://github.com/opencv/opencv/files/12926610/model1_withFlatten.zip)

Another important discovery I made is that the identical model, when omitting the flatten layer, functions correctly with OpenCV 4.8.1. This suggests that there might be an issue specifically related to the flatten layer.

Both models:
[model1.zip](https://github.com/opencv/opencv/files/12926672/model1.zip)
[model1_withFlatten.zip](https://github.com/opencv/opencv/files/12926673/model1_withFlatten.zip)

Thank you in advance!

### Steps to reproduce

```
#include <iostream>
#include <filesystem>
#include <chrono>
#include <thread>
#include <fstream>

#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/dnn.hpp>
#include <opencv2/core.hpp>

int main()
{
    unsigned int num_inferences = 100;
    cv::Size inputSizeModel1 = cv::Size(112, 112);
    int numChannels = 3;
    std::string imageFilename = "C:/Users/cesar.gouveia/Downloads/test.jpeg";
    std::string model1ONNX = "C:/Users/cesar.gouveia/Downloads/model1.onnx";
    std::string model1ONNX_withFlatten = "C:/Users/cesar.gouveia/Downloads/model1_withFlatten.onnx";

    cv::dnn::Net net = cv::dnn::readNetFromONNX(model1ONNX);
    net.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
    net.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);

     cv::Mat img = cv::imread(imageFilename);

    cv::Mat resized;
    cv::resize(img, resized, inputSizeModel1);

    std::cout << "Image size: " << resized.channels() << "x" << resized.size[0] << "x" << resized.size[1] << std::endl;
    std::cout << "Image depth: " << resized.depth() << std::endl;

    std::vector<cv::Mat> imgBatch = { resized };
    bool swapRBChannels = false;
    cv::Mat blob = cv::dnn::blobFromImages(imgBatch, 1.0, cv::Size(), cv::Scalar(), swapRBChannels, false, CV_32F);

    std::cout << "Blob size: " << blob.size[0] << "x" << blob.size[1] << "x" << blob.size[2] << "x" << blob.size[3] << std::endl;

    net.setInput(blob);

    std::vector<cv::String> unconnectedOutLayerNames = net.getUnconnectedOutLayersNames();

    std::vector<cv::Mat> outputs;
    outputs.clear();

    std::chrono::steady_clock::time_point timeLoadModelPlusInference1 = std::chrono::high_resolution_clock::now();

    net.forward(outputs, unconnectedOutLayerNames);

    std::chrono::steady_clock::time_point timeLoadModelPlusInference2 = std::chrono::high_resolution_clock::now();

    std::chrono::duration<double, std::milli> ms_doubleTimeLoadModelPlusInference = timeLoadModelPlusInference2 - timeLoadModelPlusInference1;

    std::cout << "Execution time (load model + inference): " << ms_doubleTimeLoadModelPlusInference.count() << std::endl; // in ms

    auto time1 = std::chrono::high_resolution_clock::now();

    try
    {
        for (size_t i = 0; i < num_inferences; i++)
            net.forward(outputs, unconnectedOutLayerNames);
    }
    catch (std::exception& ex)
    {
        std::cout << ex.what() << std::endl;
    }
    auto time2 = std::chrono::high_resolution_clock::now();

    std::chrono::duration<double, std::milli> ms_double = time2 - time1;

    std::cout << "Execution time inference only: " << ms_double.count() / num_inferences << std::endl; // in ms

    std::cout << outputs[0].size[0] << std::endl;
    //std::cout << outputs[0].size[1] << std::endl;
    //std::cout << outputs[0].size[2] << std::endl;
    std::cout << "Finished" << std::endl;
}

```

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

**Comment by asmorkalov** [comment]
Confirmed with code:
Code:
```
#include <iostream>
#include <chrono>
#include <thread>
#include <fstream>

#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/dnn.hpp>
#include <opencv2/core.hpp>

int main()
{
    unsigned int num_inferences = 100;
    cv::Size inputSizeModel1 = cv::Size(112, 112);
    int numChannels = 3;
    std::string imageFilename = "lena.jpg";
    std::string model1ONNX = "model1.onnx";
    std::string model1ONNX_withFlatten = "model1_withFlatten.onnx";

    cv::dnn::Net net = cv::dnn::readNetFromONNX(model1ONNX);
    net.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
    net.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);

     cv::Mat img = cv::imread(imageFilename);

    cv::Mat resized;
    cv::resize(img, resized, inputSizeModel1);

    std::cout << "Image size: " << resized.channels() << "x" << resized.size[0] << "x" << resized.size[1] << std::endl;
    std::cout << "Image depth: " << resized.depth() << std::endl;

    std::vector<cv::Mat> imgBatch = { resized };
    bool swapRBChannels = false;
    cv::Mat blob = cv::dnn::blobFromImages(imgBatch, 1.0, cv::Size(), cv::Scalar(), swapRBChannels, false, CV_32F);

    std::cout << "Blob size: " << blob.size[0] << "x" << blob.size[1] << "x" << blob.size[2] << "x" << blob.size[3] << std::endl;

    net.setInput(blob);

    std::vector<cv::String> unconnectedOutLayerNames = net.getUnconnectedOutLayersNames();

    std::vector<cv::Mat> outputs;
    outputs.clear();

    net.forward(outputs, unconnectedOutLayerNames);

    std::cout << "Finished 1" << std::endl;

    try
    {
        for (size_t i = 0; i < num_inferences; i++)
            net.forward(outputs, unconnectedOutLayerNames);
    }
    catch (std::exception& ex)
    {
        std::cout << ex.what() << std::endl;
    }

    std::cout << "Outputs: " << outputs[0].size[0] << std::endl;
    //std::cout << outputs[0].size[1] << std::endl;
    //std::cout << outputs[0].size[2] << std::endl;
    std::cout << "Finished 2" << std::endl;
}
```
Output:
```
[ERROR:0@0.025] global onnx_importer.cpp:1032 handleNode DNN/ONNX: ERROR during processing node with 3 inputs and 1 outputs: [Gemm]:(onnx_node!pre_fc1) from domain='ai.onnx'
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.8.0-dev) /mnt/projects/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_importer.cpp:1051: error: (-2:Unspecified error) in function 'handleNode'
> Node [Gemm@ai.onnx]:(onnx_node!pre_fc1) parse error: OpenCV(4.8.0-dev) /mnt/projects/Projects/OpenCV/opencv-master/modules/dnn/src/layers/gemm_layer.cpp:70: error: (-2:Unspecified error) in function 'virtual bool cv::dnn::GemmLayerImpl::getMemoryShapes(const std::vector<std::vector<int> >&, int, std::vector<std::vector<int> >&, std::vector<std::vector<int> >&) const'
> > DNN/Gemm: Invalid dimension of dim K (expected: 'K_a == K_b'), where
> >     'K_a' is 1
> > must be equal to
> >     'K_b' is 512
> 
Segmentation fault (core dumped)
```

**Comment by asmorkalov** [comment]
@fengyuentau Could you take a look?

**Comment by fengyuentau** [comment]
@cesarpgouveia Thank you for the report. To clarify, you provide three models and I will explain the situation one by one.

1. https://github.com/opencv/opencv/files/12926610/model1_withFlatten.zip. I guess you name it wrong with this model because it does not have flatten in it, and it is not valid since `Gemm` operator requires 2d tensors as inputs.
2. https://github.com/opencv/opencv/files/12926672/model1.zip. This one works fine with the latest opencv 4.x code.
3. https://github.com/opencv/opencv/files/12926673/model1_withFlatten.zip. This one has the same SHA encoding with the first one, so they are the same.

Also note that all the three models you provided are actually invalid for ONNX Runtime because maybe you manually modified the model to have the `Sub` and `Mul` operators at the beginning, which have input B of data type `float64`, while input A data type is `float32`. They need to be the same data type.

**Comment by cesarpgouveia** [comment]
Yes, sorry, the naming is swapped, meaning that "model1" is the model that contains the flattening layer, and "model1_withFlatten" is the model without the flattening layer. Here are the models again with the correct naming:

[models.zip](https://github.com/opencv/opencv/files/13004043/models.zip)

and yes, it's the one without the flatten layer that actually has the problem.

> Also note that all three models you provided are actually invalid for ONNX Runtime because maybe you manually modified the model to have the Sub and Mul operators at the beginning, which have input B of data type float64, while input A data type is float32. They need to be the same data type.
Is this a new thing in OpenCV 4.8.*? because I have at least 3 models with that kind of logic at the beginning and they were working fine with OpenCV 4.7.*.

> and it is not valid since Gemm operator requires 2d tensors as inputs.

But it was valid on OpenCV 4.7.*? I just want to understand what changed and how can I modify my models to accommodate OpenCV 4.8.* on my solution.

Thank you for your help!

**Comment by cesarpgouveia** [comment]
**Regarding the matter of inputs A and B having different data types,** I just used this python script to convert from float64 (double) to float32 (float):

```
import onnx
import onnx.helper
import onnx.shape_inference

# Load the ONNX model
onnx_model = onnx.load(r"C:\Users\cesar.gouveia\Downloads\model1\modified_model.onnx")

# Identify the name of the weight tensor you want to change
target_weight_name = "scalar_op1"

# Find the target weight tensor
target_weight = None
for initializer in onnx_model.graph.initializer:
    if initializer.name == target_weight_name:
        target_weight = initializer
        break

print(target_weight)

# Check if the target weight was found
if target_weight is not None:
    # Create a new initializer with the desired data type and dimensions
    new_data_type = onnx.TensorProto.FLOAT  # Change this to your desired data type
    new_dims = target_weight.dims  # Keep the same dimensions

    new_weight = onnx.helper.make_tensor(
        name=target_weight_name,
        data_type=new_data_type,
        dims=new_dims,
        vals=[float(target_weight.double_data[0])]
    )

    # Remove the old initializer from the graph
    onnx_model.graph.initializer.remove(target_weight)

    # Append the new initializer to the graph's initializers
    onnx_model.graph.initializer.append(new_weight)

    # Update shape information if necessary
    #onnx_model = onnx.shape_inference.infer_shapes(onnx_model)

    # Save the modified ONNX model
    onnx.save(onnx_model, r"C:\Users\cesar.gouveia\Downloads\model1\modified_model_both.onnx")
else:
    print(f"Weight tensor with name '{target_weight_name}' not found in the model.")
```

Now I have "model1.onnx" which uses float32 for input A and float64 for inputB, and "model1_modified_float32.onnx" which uses float32 for sub and mul operations and both inputA and inputB of those operations. However, they are compatible with OpenCV, regardless of whether the OpenCV version is 4.7 or 4.8 or higher. So my question is why do you say that ONNX Runtime has issues with different data types (at least for float32 and float64 - double). The models are going to yield different results? If so, is the disparity between the models significant?

**Comment by cesarpgouveia** [comment]
Regarding the Gemm Layer Issue: I successfully resolved the problem with my model by introducing a Flatten Layer prior to the Gemm Layer. For those encountering the same issue, I utilized the following code to insert the Flatten Layer between the convolutional and Gemm layers. 

```
import onnx
from onnx import helper
from onnx import AttributeProto
from onnx import shape_inference

def add_flatten_before_gemm(model_path, output_model_path):
    # Load the original ONNX model
    onnx_model = onnx.load(model_path)

    # inferred_model = shape_inference.infer_shapes(onnx_model)
    # print(inferred_model.graph.value_info)

    # Define a Flatten operator
    flatten_node = onnx.helper.make_node(
        op_type='Flatten',
        inputs=['conv_6dw7_7_batchnorm'],  # Replace 'input' with the actual input name
        outputs=['flatten_426'],  # Replace 'flattened_output' with the desired output name
        #axis=1  # Adjust the axis based on your model's input shape
    )

    # Iterate through the model's nodes and identify GEMM layers
    for i, node in enumerate(onnx_model.graph.node):
        if node.op_type == 'Gemm':
            # Insert a Flatten node before the GEMM layer
            flatten_node.input[0] = node.input[0]
            flatten_node.output[0] = node.input[0] + '_flattened'
            onnx_model.graph.node.insert(i, flatten_node)
            node.input[0] = flatten_node.output[0] # then the input of the next layer has to be the same name as the output of the layer we just created
            break

    # Save the modified ONNX model
    onnx.save(onnx_model, output_model_path)


if __name__ == "__main__":
    input_model_path = r"C:\Users\cesar.gouveia\Projects\vbfacematchers\model1.onnx"
    output_model_path = r"C:\Users\cesar.gouveia\Projects\vbfacematchers\model1_with_flatten.onnx"
    add_flatten_before_gemm(input_model_path, output_model_path)
```

This adjustment rectifies the problem on OpenCV version 4.8 or newer. Nevertheless, it's worth noting that in the previous version of OpenCV (4.7 or older), no Flatten Layer was required.

Once again, thank you [fengyuentau](https://github.com/fengyuentau) for identifying these issues, and I remain eager to understand why this problem arises in the new version of OpenCV.

**Comment by fengyuentau** [comment]
Okay, I will try to reply to you one by one:

---

Your new attached models https://github.com/opencv/opencv/files/13004043/models.zip. The flatten one should work correctly.

---

> Is this a new thing in OpenCV 4.8.+?

Yes, but also this is the specification of ONNX. In previous dnn version, we do not consider that strictly, but now we will try to follow the spec as possible as we can. See [Gemm spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Gemm) (it requires both inputs are 2d tensor) and [Sub spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md#Sub) (it needs both inputs of the same data type). This is why I DO NOT recommand manual modifications on ONNX models, which may easily break them unless you exactly know what you are doing.

---

> But it was valid on OpenCV 4.7.+?

Again see the Gemm spec from ONNX documentation.

---

> So my question is why do you say that ONNX Runtime has issues with different data types (at least for float32 and float64 - double). The models are going to yield different results? If so, is the disparity between the models significant?

1. Just have a try running your previous model with ONNX Runtime and it simply complains about this issue.
2. In common cases, they are not going to yield different results. This is the spec and a normal ONNX exporter (such as `torch.onnx.export`) strictly follows the spec.

---

> Nevertheless, it's worth noting that in the previous version of OpenCV (4.7 or older), no Flatten Layer was required.

Also note that in 4.8.1 we reworked Gemm and it runs significantly faster on CPU if you use opencv implemented backend. Details see https://github.com/opencv/opencv/pull/23897. Also we are going to optimize the n-dimensional MatMul (works like [np.matmul](https://numpy.org/doc/stable/reference/generated/numpy.matmul.html)), which needs a correct and success shape inference. We are getting more closer to the ONNX spec, so please ensure your ONNX model is valid (use `onnx.checker.check_model` and `onnxruntime` to check).

**Comment by cesarpgouveia** [comment]
Thanks for the explanation @fengyuentau!

> Yes, but also this is the specification of ONNX. In previous dnn version, we do not consider that strictly, but now we will try to follow the spec as possible as we can.

Got it!

> Also note that in 4.8.1 we reworked Gemm and it runs significantly faster on CPU if you use opencv implemented backend.

Faster than the OpenVINO backend?

> We are getting more closer to the ONNX spec, so please ensure your ONNX model is valid (use onnx.checker.check_model(model_proto, full_check=True) to check

I was unaware of this, this is important!

I believe we can now resolve this issue, as @fengyuentau has clarified that the problem does not stem from the new OpenCV code (OpenCV 4.8+), but rather from my model architectures, which were incompatible with ONNX Runtime."

**Comment by fengyuentau** [comment]
> Faster than the OpenVINO backend?

OpenVINO is optimized specifically for Intel CPUs. They have their own magic for optimization. The purpose for the reworking is models can run faster on different platforms including ARM, RISC-V and so on.

I will close this for now. Feel free to reopen if you are still confused.
